### PR TITLE
Add to_table() to RetrievalJob object

### PR DESCRIPTION
**Pull request by MattDelac**
**What this PR does / why we need it**:
We should add a notion of "OfflineJob" so that it will be easier to add new methods as well as accessible useful information.

For example, when you build a custom Provider, you might want to access the query coming from `pull_latest_from_table_or_query()` without executing it (as it will be executed differently)

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
pull_latest_from_table_or_query() now returns an RetrievalJob object
Add method to_table() to RetrievalJob object
```

### CODE REVIEW

File: sdk/python/feast/infra/offline_stores/bigquery.py [hunk 1]
```diff
@@ -228,6 +225,15 @@ def _infer_event_timestamp_from_dataframe(entity_df: pandas.DataFrame) -> str:
             )
 
 
+class BigQueryOfflineJob(OfflineJob):
+    def __init__(self, query: str, config: RepoConfig):
+        self.query = query
+        self.client = _get_bigquery_client(project=config.offline_store.project_id)
+
+    def to_table(self) -> pyarrow.Table:
+        return self.client.query(self.query).to_arrow()
```
Comment (woop) [inline sdk/python/feast/infra/offline_stores/bigquery.py:234]: Is it intentional for `BigQueryOfflineJob.to_arrow()` to return a pyarrow table? I would have expected `to_table()` to create a BQ table. Would `to_arrow_table()` or `to_arrow()` be clearer?

**Comment by woop** [comment]
I really like this idea @MattDelac!

The one jarring thing is that we now have two types of jobs that are completely isolated. `RetrievalJob` and `OfflineJob`. Is there any reason for these to be different?

I like the idea of having a single `OfflineJob` that is configured differently through methods like `materialize` and `get_historical_features()` and context objects, and then you automatically get all the `to_sql` and other base functionality. The wrapping command (`materialize`, `get_historical_features`) can then work with the OfflineJob differently, like in one case it can use `.to_table()` or whatever if it needs to load the latest values into the online store.

What do you think about that?

**Comment by MattDelac** [comment]
> The one jarring thing is that we now have two types of jobs that are completely isolated. `RetrievalJob` and `OfflineJob`. Is there any reason for these to be different?

Hmmm good question 🤔 

I can see if those two objects could/should be the same

**Comment by MattDelac** [comment]
@woop I changed this PR to add a new method to the `RetrievalJob` object instead of creating a new object

**Comment by woop** [comment]
> @woop I changed this PR to add a new method to the `RetrievalJob` object instead of creating a new object

Are you ready for a review?

**Comment by MattDelac** [comment]
> > @woop I changed this PR to add a new method to the `RetrievalJob` object instead of creating a new object
> 
> Are you ready for a review?

Let me check this linter issue since I rebased on master

**Comment by MattDelac** [comment]
> > @woop I changed this PR to add a new method to the `RetrievalJob` object instead of creating a new object
> 
> Are you ready for a review?

It sounds good now 👍

**Review by achals — COMMENTED** [review]
looks good to me. Can we exercise this code path in the existing integration tests?

**Comment by MattDelac** [comment]
> Can we exercise this code path in the existing integration tests?

Not sure I understand what you mean here ...

**Comment by achals** [comment]
> > Can we exercise this code path in the existing integration tests?
> 
> Not sure I understand what you mean here ...

Sorry, i meant can we call `to_table` during one of the integration tests?

**Comment by woop** [comment]
Also lgtm

**Review by achals — APPROVED** [review]
/lgtm

**Comment by woop** [comment]
@achals @MattDelac I know we merged this, but I just realized that this comment was unaddressed https://github.com/feast-dev/feast/pull/1663#discussion_r659032133

How do you feel about making the method `to_arrow()`?

**Comment by MattDelac** [comment]
> @achals @MattDelac I know we merged this, but I just realized that this comment was unaddressed [#1663 (comment)](https://github.com/feast-dev/feast/pull/1663#discussion_r659032133)
> 
> How do you feel about making the method `to_arrow()`?
> 
> Example in the wild https://googleapis.dev/python/bigquerystorage/latest/bigquery_storage_v1/library.html?highlight=arrow#google.cloud.bigquery_storage_v1.reader.ReadRowsIterable.to_arrow

Ho sorry it's my fault. Will push a quick fix where I rename it `to_arrow()` 👍

**Pull request merged**